### PR TITLE
ShortCircuit sets up chunked transfer automatically when necessary

### PR DESCRIPTION
**Pull request by oxtoacart**
For getlantern/lantern-internal#1413.

### CODE REVIEW

File: filters/filter.go [hunk 1]
```diff
@@ -32,6 +32,10 @@ func ShortCircuit(ctx Context, req *http.Request, resp *http.Response) (*http.Re
 	resp.Proto = req.Proto
 	resp.ProtoMajor = req.ProtoMajor
 	resp.ProtoMinor = req.ProtoMinor
+	if resp.Body != nil && resp.ContentLength <= 0 && len(resp.TransferEncoding) == 0 {
```
Comment (myleshorton) [inline filters/filter.go]: I guess the part I'm confused about here @oxtoacart is that chunked encoding requires actual chunking in the body with chunk delimiters and such right? Are you seeing that in practice when you see this combination the body is, in fact, chunked with the default delimiter or something? I may be misremembering how the delimiter works exactly too.

Comment (oxtoacart) [reply filters/filter.go]: This is a response that we created. The chunking happens when we write it our with [`response.Write()`](https://golang.org/pkg/net/http/#Response.Write). Under the covers, this uses a [`transfer_writer`](https://golang.org/src/net/http/transfer.go) which handles the chunking stuff.

The problem prior to this fix is that the transfer writer is writing this like a non-chunked response, but we don't have a content-length header so the client just hangs after reading the body. By setting a negative content length and specifying chunked encoding, we get the transfer writer to actually transfer it as chunked and send the right transfer-encoding header.

Comment (oxtoacart) [reply filters/filter.go]: The way I tested this was by teeing the written data to `stderr` and making sure it looked chunked, and also by running `checkfallbacks` against this and seeing that it no longer hangs.

Comment (myleshorton) [reply filters/filter.go]: OK cool. The only thing I'm wondering about is content lengths of 0. Presumably they would not have a body, so should be fine in theory, but I'd be tempted to change it to `resp.ContentLength < 0`. Even though 0 is the default value for int64, AFAICT the code always sets it to -1 if there actually is no content length.

Comment (oxtoacart) [reply filters/filter.go]: > AFAICT the code always sets it to -1 if there actually is no content length.

`ShortCircuit` can be (and is) called from a variety of places. We don't really have control over what they set as the Content Length. This logic basically says "hey, we're an HTTP 1.1 proxy, so we can chunk. If there's a body, and you didn't give us neither a real Content Length nor specify your own transfer encoding, we're going to chunk."

Comment (myleshorton) [reply filters/filter.go]: OK works for me...

**Comment by myleshorton** [comment]
LGTM!

**Pull request merged**